Write Service Status Life Cycle test - +4 endpoint coverage

riaankleinhans opened the pull request:
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it:**
This PR adds a test to test the following untested endpoints:
- patchCoreV1NamespacedServiceStatus
- readCoreV1NamespacedServiceStatus
- replaceCoreV1NamespacedServiceStatus
- patchCoreV1NamespacedService

**Which issue(s) this PR fixes:**
Fixes #94867
**Testgrid Link:**
[Link](https://testgrid.k8s.io/sig-release-master-blocking#gce-cos-master-default&include-filter-by-regex=should.complete.a.service.status.lifecycle&width=5&graph-metrics=test-duration-minutes)

**Special notes for your reviewer:**
Adds +4 endpoint test coverage (good for conformance)

**Does this PR introduce a user-facing change?:**
```
NONE

```

**Release note:**
```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:**
```
NONE

```

/sig testing
/sig architecture
/area conformance

Riaankl commented:
/triage accepted

Riaankl commented:
/assign @smarterclayton 
This is from the discussion in the 12 Jan [Conformance meeting](https://github.com/kubernetes/kubernetes/issues/94867#issuecomment-759016564)

Riaankl commented:
/retest

Riaankl commented:
/retest

Riaankl commented:
/retest

Riaankl commented:
/test all

heyste commented:
/test pull-kubernetes-e2e-ubuntu-gce-network-policies
flake `Kubernetes  e2e suite: [sig-network] Networking IPerf [Experimental] [Slow]  [Feature:Networking-Performance] should transfer ~ 1GB onto the service  endpoint 1 servers (maximum of 1 clients)`

```
I0113 23:52:52.604] Jan 13 23:52:44.588: INFO: Selector matched 1 pods for map[app:iperf-e2e-cli-pod]
I0113 23:52:52.604] Jan 13 23:52:44.589: INFO: ForEach: Found 1 pods from the filter.  Now looping through them.
I0113 23:52:52.604] Jan 13 23:52:44.589: INFO: Running '/go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/kubectl --server=https://35.233.212.159 --kubeconfig=/workspace/.kube/config --namespace=network-perf-2835 logs iperf-e2e-cli-pod-0 iperf-client'
I0113 23:52:52.604] Jan 13 23:52:44.903: INFO: stderr: ""
I0113 23:52:52.604] Jan 13 23:52:44.903: INFO: stdout: "connect failed: Connection refused\n"
I0113 23:52:52.605] Jan 13 23:52:46.904: FAIL: Unexpected error, Failed to find "0-", last result: "connect failed: Connection refused
I0113 23:52:52.605] " when running forEach on the pods. 
```

Riaankl commented:
/test pull-kubernetes-e2e-ubuntu-gce-network-policies

`I0114 01:20:42.031] [Fail] [sig-network] Networking IPerf [Experimental] [Slow] [Feature:Networking-Performance] [It] should transfer ~ 1GB onto the service endpoint 1 servers (maximum of 1 clients) 
I0114 01:20:42.031] test/e2e/framework/framework.go:809
I0114 01:20:42.031] 
I0114 01:20:42.031] Ran 405 of 5705 Specs in 1676.663 seconds
I0114 01:20:42.031] FAIL! -- 404 Passed | 1 Failed | 0 Pending | 5300 Skipped`

JornShen commented on test/e2e/network/service.go:
```diff
@@ -2947,6 +2948,118 @@ var _ = SIGDescribe("Services", func() {
+		}
+		framework.ExpectEqual(foundSvc, true, "unable to find Service in list of Services.")
+
+		ginkgo.By("deleting the service")
```
suggest move this  after create service(line 2976). use `defer` to confirm service will be removed whether processe failed after service  
created.

heyste replied:
After the test runs the e2e framework cleans up any resources especially those in the test namespace

JornShen replied:
which place? @heyste 

this place?

https://github.com/kubernetes/kubernetes/blob/b3dd01da497f4f9f81ffb2ebf275ff08cbfebaa2/test/e2e/network/service.go#L744-L754

JornShen replied:
leaving it for e2e framework clean is also ok. :)

JornShen commented on test/e2e/network/service.go:
```diff
@@ -2947,6 +2948,118 @@ var _ = SIGDescribe("Services", func() {
+		dc := f.DynamicClient
+		ns := f.Namespace.Name
+		svcResource := schema.GroupVersionResource{Group: "", Version: "v1", Resource: "services"}
+		testServiceName := "test-service"
```
use a random code as part of name?

heyste replied:
Fixed by using `"k8s.io/apimachinery/pkg/util/rand"` package
```
testSvcName := "test-service-" + utilrand.String(5)
```

Riaankl commented:
/test all

JornShen commented on test/e2e/network/service.go:
```diff
@@ -2947,6 +2948,118 @@ var _ = SIGDescribe("Services", func() {
+		framework.ExpectNoError(err, "failed to create service")
+
+		svcWatchChan := svcWatch.ResultChan()
+		for event := range svcWatchChan {
```
what if  status is still pending(adding)?  will this test no exist?

heyste replied:
After chatting with @BobyMCbobs I've updated the test to use some better processes around watches. 

@JornShen Does the current test process address your concerns?

JornShen replied:
@heyste yes, impl is good.

Riaankl commented:
/test all

Riaankl commented:
/test all

Riaankl commented:
/test all

Riaankl commented:
/test all

Riaankl commented:
@JornShen Thank you for your thorough review of the PR. 
Did @heyste manage to address all your suggestions satisfactorily?
If it is good to go we would like to get this merged to give it sufficient time on the test grid to ensure a solid test for this release.

JornShen commented:
/test pull-kubernetes-e2e-ubuntu-gce-network-policies

JornShen commented:
a flaky test 🙃 
```
test/e2e/network/conntrack.go:129
Jan 28 02:39:28.652: Failed to connect to backend 1
vendor/github.com/onsi/ginkgo/internal/leafnodes/runner.go:113
```

let's give another test 

/test pull-kubernetes-e2e-ubuntu-gce-network-policies

JornShen commented:
/lgtm

/cc @aojea

aojea commented on test/e2e/network/service.go:
```diff
@@ -2947,6 +2951,179 @@ var _ = SIGDescribe("Services", func() {
+		framework.ExpectNoError(err, "failed to locate Service %v in namespace %v", testService.ObjectMeta.Name, ns)
+		framework.Logf("Service %s created", testSvcName)
+
+		ginkgo.By("patching the ServiceStatus")
+		serviceStatusPatch, err := json.Marshal(map[string]interface{}{
+			"metadata": map[string]interface{}{
+				"labels": map[string]string{"test-service": "patched"},
+			},
+			"spec": map[string]interface{}{
+				"ports": []map[string]interface{}{{
+					"name":       "http8080",
+					"port":       int32(8080),
+					"targetPort": int(8080),
+					"selector": []map[string]interface{}{{
+						"type": "LoadBalancer",
+					}},
+				}},
+			},
+		})
```
don´t you have to patch the Status too?
https://kubernetes.io/docs/reference/kubernetes-api/services-resources/service-v1/#ServiceStatus

```
status:
  loadBalancer:
    ingress:
    - ip: 203.0.113.1
  ```

heyste replied:
> don´t you have to patch the Status too?
> https://kubernetes.io/docs/reference/kubernetes-api/services-resources/service-v1/#ServiceStatus
> 
> ```
> status:
>   loadBalancer:
>     ingress:
>     - ip: 203.0.113.1
> ```

When tracking this test with apisnoop the test does hit the `patchCoreV1NamespacedServiceStatus` endpoint. The test approach was reviewed with @smarterclayton in the [Jan 12 conformance meeting](https://github.com/kubernetes/kubernetes/issues/94867#issuecomment-759016564).

johnbelamaric replied:
that's confusing to me as well. do we have an explanation why this test hits status? I mean, we list the status sub-resource, but we don't list any changes to it. @smarterclayton is that sufficient to say it works?

heyste replied:
I asked Clayton in that [conformance meeting](https://www.youtube.com/watch?v=ENXq4UyQr9I&t=488) if the test process was on the right track:
>  If you are patching status you may mutate the both status and metadata even though it it is s a status endpoint there are historical reasons for that and we couldn't redefine it. And in the spec the  full object should be returned, but only mutations in status and metadata should be accepted. @smarterclayton 

See also mailing list message from @hh https://groups.google.com/g/kubernetes-sig-architecture/c/Ekv2NF2D5B8

johnbelamaric replied:
But that says: "you may mutate both status and metadata", but the problem is we are *not* mutating status. I think we should be.

heyste replied:
This is the audit log entry from [pull-kubernetes-e2e-gce-ubuntu-containerd #1352299357109489664](https://prow.k8s.io/view/gcs/kubernetes-jenkins/pr-logs/pull/98018/pull-kubernetes-e2e-gce-ubuntu-containerd/1352299357109489664/) that shows the patching of the endpoint.

```
curl -L \
https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/98018/pull-kubernetes-e2e-gce-ubuntu-containerd/1352299357109489664/artifacts/e2e-5d7b0f355d-a7d53-master/kube-apiserver-audit.log-20210121-1611249901.gz  \
| gzip -d   \
| grep "should complete a service status lifecycle" \
| grep patch \
| jq . 
```

```json
{
"requestURI": "/api/v1/namespaces/services-8164/services/test-service-szbkj/status",
  "verb": "patch",
 "userAgent": "e2e.test/v1.21.0 (linux/amd64) kubernetes/4405248 -- \
[sig-network] Services should complete a service status lifecycle",
   "responseStatus": {
    "metadata": {},
    "code": 200
  },
 "requestObject": {
    "metadata": {
      "labels": {
        "test-service": "patched"
      }
    },
    "spec": {
      "ports": [
        {
          "name": "http8080",
          "port": 8080,
          "selector": [
            {
              "type": "LoadBalancer"
            }
          ],
          "targetPort": 8080
        }
      ]
    }
  }
}
```
```

johnbelamaric replied:
Sure, but it doesn't actually *mutate* the status. It just mutates the metadata. So what we're really testing is "can I mutate metadata via the status endpoint" which is nice and all but it's not the primary purpose of the status endpoint.

Is there any field in the status we can mutate that won't cause a race with the controller?

johnbelamaric replied:
After a brief convo with @liggitt I realized that since we now use watch, we shouldn't have an issue with races with the controller, since we will see our change (even if we do see the controller change it back *after* that, we can ignore it).

so, you should be patching a specific field in status. just pick one...

heyste replied:
What if we create an arbitrary `status.condition` ?
```
# kubectl explain service.status.conditions
KIND:     Service
VERSION:  v1

RESOURCE: conditions <[]Object>

DESCRIPTION:
     Current service state

     Condition contains details for one aspect of the current state of this API
     Resource.

FIELDS:
   lastTransitionTime	<string> -required-
     lastTransitionTime is the last time the condition transitioned from one
     status to another. This should be when the underlying condition changed. If
     that is not known, then using the time when the API field changed is
     acceptable.

   message	<string> -required-
     message is a human readable message indicating details about the
     transition. This may be an empty string.

   observedGeneration	<integer>
     observedGeneration represents the .metadata.generation that the condition
     was set based upon. For instance, if .metadata.generation is currently 12,
     but the .status.conditions[x].observedGeneration is 9, the condition is out
     of date with respect to the current state of the instance.

   reason	<string> -required-
     reason contains a programmatic identifier indicating the reason for the
     condition's last transition. Producers of specific condition types may
     define expected values and meanings for this field, and whether the values
     are considered a guaranteed API. The value should be a CamelCase string.
     This field may not be empty.

   status	<string> -required-
     status of the condition, one of True, False, Unknown.

   type	<string> -required-
     type of condition in CamelCase or in foo.example.com/CamelCase.
```

aojea replied:
this is the more realistic , I guess
```
status:
  loadBalancer:
    ingress:
    - ip: 203.0.113.1
```

dims replied:
Looks like this is taken care of as well.

```
+		ginkgo.By("patching the ServiceStatus")
+		lbStatus := v1.LoadBalancerStatus{
+			Ingress: []v1.LoadBalancerIngress{{IP: "203.0.113.1"}},
+		}
```

Riaankl commented:
@smarterclayton look like all is OK for this PR to Merge.
Would you please add an `/approve` if you are ok with it?

Riaankl commented:
/assign @johnbelamaric

heyste commented on test/e2e/network/service.go:
```diff
@@ -2947,6 +2951,179 @@ var _ = SIGDescribe("Services", func() {
+			},
+		})
+		framework.ExpectNoError(err, "Could not marshal json", err)
+		_, err = dc.Resource(svcResource).Namespace(ns).Patch(context.TODO(), testSvcName, types.StrategicMergePatchType, []byte(serviceStatusPatch), metav1.PatchOptions{}, "status")
```
@johnbelamaric note the last parameter of the patch function is the "status" subresource.

johnbelamaric replied:
yes, but the content of the patch doesn't touch status

dims replied:
looks like this is fixed

```
+		framework.ExpectNoError(err, "Failed to marshal JSON. %v", err)
+		_, err = svcClient.Patch(context.TODO(), testSvcName, types.MergePatchType,
+			[]byte(`{"metadata":{"annotations":{"patchedstatus":"true"}},"status":{"loadBalancer":`+string(lbStatusJSON)+`}}`),
+			metav1.PatchOptions{}, "status")
+		framework.ExpectNoError(err, "Could not patch service status", err)
```

heyste commented:
/test pull-kubernetes-e2e-kind
flake

```
Kubernetes e2e suite: [sig-network] EndpointSlice should create Endpoints and EndpointSlices for Pods matching a Service  3m1s

test/e2e/network/endpointslice.go:167 
Feb 18 01:43:36.146: timed out waiting for Pods to have IPs assigned
Unexpected error:
    <*errors.errorString \| 0xc0001f81d0>: {
        s: "timed out waiting for the condition",
    }
    timed out waiting for the condition
occurred
test/e2e/network/endpointslice.go:289
```

dims commented:
@johnbelamaric all the comments have been addressed. please take a look. Do you want any other sig-network approvers to chime in too?

Riaankl commented:
@heyste have updated the `Status` testing as discussed and we believe the test is good to go on the test grid. There are only 3 weeks left before test freeze 1.21, and we need 2 weeks flake free before Promoting.
Can you please review @johnbelamaric @andrewsykim

heyste committed:
Add lifecycle e2e test for service status

e2e test validates the following service status endpoints
- patchCoreV1NamespacedServiceStatus
- readCoreV1NamespacedServiceStatus
- replaceCoreV1NamespacedServiceStatus

Also includes untested service endpoint
- patchCoreV1NamespacedService

aojea commented:
I don't want to block on this comment, just want to open a discussion based on this e2e test approach for following up later.

I'm far from an expert on the area, but I always see that the recommendation is to not use watchers directly because they are not able to handle disconnects, most of the e2e use the `wait.PollImmediate()` method for checking the resources and synchronizing the steps with that active loop.

@BenTheElder @wojtek-t @liggitt what do you think? I think that it will be useful to set some guidelines on the e2e tests so we have some consistency on the code.

dims commented on test/e2e/network/service.go:
```diff
@@ -2199,6 +2204,223 @@ var _ = common.SIGDescribe("Services", func() {
+			Ingress: []v1.LoadBalancerIngress{{IP: "203.0.113.1"}},
+		}
+		lbStatusJSON, err := json.Marshal(lbStatus)
+		framework.ExpectNoError(err, "Failed to marshal JSON. %v", err)
```
this looks funny :) `err` twice

dims replied:
just a nit. no need to fix right now

dims commented:
/approve
/lgtm

let's please iterate if needed. we need to get this to soak quickly

cc @spiffxp @johnbelamaric @liggitt

johnbelamaric commented:
/lgtm
/approve

@aojea we have built some defenses in for watch failures, we do want to use watch in a number of these tests. in this one in particular, we need to use watch otherwise we may miss our update due to fighting with the controller that is actually managing the resource.

aojea commented:
👍 cool, nice to know

Riaankl commented:
/test pull-kubernetes-e2e-kind

Unrelated flake

```
Kubernetes e2e suite: [sig-storage] CSI Volumes [Driver: csi-hostpath] [Testpattern: Dynamic PV (default fs)] volumes should store data expand_more	6m7s
Kubernetes e2e suite: [sig-cli] Kubectl client Simple pod should return command exit codes expand_more	4m38s
Kubernetes e2e suite: [sig-storage] In-tree Volumes [Driver: local][LocalVolumeType: block] [Testpattern: Pre-provisioned PV (ext4)] volumes should store data expand_more
```

Riaankl commented:
/test pull-kubernetes-integration

Unrelated flake

```
2/3181 Tests Failed. | expand_less
k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache/timed_out_client_requests_skip_later_mutating_webhooks_(regardless_of_failure_policy)_and_fail expand_more3s | k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache/timed_out_client_requests_skip_later_mutating_webhooks_(regardless_of_failure_policy)_and_fail expand_more | 3s
k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache/timed_out_client_requests_skip_later_mutating_webhooks_(regardless_of_failure_policy)_and_fail expand_more | 3s
k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache expand_more | k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache expand_more
k8s.io/kubernetes/test/integration/apiserver/admissionwebhook: TestWebhookTimeoutWithWatchCache expand_more
```

Riaankl commented:
/test pull-kubernetes-e2e-kind
unrelated flake
```
• Failure [180.934 seconds]
[sig-network] EndpointSlice
/home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/common/framework.go:23
  should create Endpoints and EndpointSlices for Pods matching a Service [It]
  /home/prow/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/network/endpointslice.go:168
  Mar  4 22:19:23.352: timed out waiting for Pods to have IPs assigned
      
  Unexpected error:
      <*errors.errorString | 0xc0002b6240>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  occurred
```
